Nldas surface eval

hcorson-dosch-usgs opened the pull request:
Code additions to evaluate the bias and accuracy of NLDAS surface predictions by these groupings: year, doy, season, and 2-degree temperature bins.

All key changes are in `5_evaluate.R` and `5_evaluate/src/eval_utility_fxns.R`

For the current MN run (3,569 lakes) there are 2,174 sites that had >10 dates with observations. This evaluation is for surface predictions for those 2,174 lakes.
_____________________________
@lindsayplatt - I'm tagging you for a quick review of my edits to the target workflow presented in #55, namely the edits to pulling together the [preds for eval](https://github.com/hcorson-dosch/lake-temperature-process-models/blob/NLDAS_surface_eval/5_evaluate.R#L37-L49) and my use of `tar_assert()` for that target and for the [pred-obs matching target](https://github.com/hcorson-dosch/lake-temperature-process-models/blob/NLDAS_surface_eval/5_evaluate.R#L64-L71).
  - After initially planning to make use of the predictions that had already been read in [when extracting the output](https://github.com/USGS-R/lake-temperature-process-models/blob/main/3_extract/src/process_glm_output.R#L30-L32), I realized when testing how the code scaled that it was too expensive (in terms of build time), as it required filtering that extracted output back down to the sites and dates for which we have observations. Instead, it was faster computationally to simply [read in the predictions from the exported files](https://github.com/hcorson-dosch/lake-temperature-process-models/blob/NLDAS_surface_eval/5_evaluate.R#L42-L46) for all sites for which we have observations and filter to dates w/ observations as the predictions are read in.

______________________________

@jread-usgs - I'm tagging you to review the functions in `5_evaluate/src/eval_utility_fxns.R`, the accuracy and bias metrics I've included, and the plots that are created.

The plots produced by these targets are [here](https://doimspp.sharepoint.com/:f:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Lakes/GLM/lake-temperature-process-models/NLDAS_surface_evaluation_plots/[MN_20220523](https://doimspp.sharepoint.com/:f:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Lakes/GLM/lake-temperature-process-models/NLDAS_surface_evaluation_plots/MN_20220523?csf=1&web=1&e=hnG1Ci)?csf=1&web=1&e=hnG1Ci). Here's one example:
![nldas_surface_accuracy_doy](https://user-images.githubusercontent.com/54007288/169883038-1e3a8164-2e46-4d5d-b8a7-d94b96a2791b.png)

If the existing code look good, I'd like to propose adding a bit more to the plots by either a) symbolizing the bars by `n_dates` / `n_sites` (color ramp) or b) pairing each bar plot w/ a histogram of `n_dates` and `n_sites` that share the same x-axis as the main plot, as a spot check reveals that at least some of the high rmse/bias values correspond to bins w/ few dates and/or sites:
![image](https://user-images.githubusercontent.com/54007288/169883308-a73a5e25-9718-42dc-88cc-d5a1db56ccbc.png)

There is considerable variability in the # of dates and sites per grouped bin (here DOY):
```r
> tar_load(p5_nldas_surface_accuracy_doy)
> median(p5_nldas_surface_accuracy_doy$n_dates)
[1] 178
> max(p5_nldas_surface_accuracy_doy$n_dates)
[1] 1177
> min(p5_nldas_surface_accuracy_doy$n_dates)
[1] 1
> median(p5_nldas_surface_accuracy_doy$n_sites)
[1] 138.5
> max(p5_nldas_surface_accuracy_doy$n_sites)
[1] 780
> min(p5_nldas_surface_accuracy_doy$n_sites)
[1] 1
```

hcorson-dosch-usgs committed:
Make name of time variable consistent

hcorson-dosch-usgs committed:
bring in temp observations

hcorson-dosch-usgs committed:
Split results extraction into combining and writing steps

hcorson-dosch-usgs committed:
Add code to generate NLDAS model evaluation targets

hcorson-dosch-usgs committed:
Switch to data.table filtering by key

hcorson-dosch-usgs committed:
Remove filter and instead make munging to long conditional

hcorson-dosch-usgs committed:
Add site_id when extracting output

hcorson-dosch-usgs committed:
Munge long w/i matching function

hcorson-dosch-usgs committed:
Filter without mapping

hcorson-dosch-usgs committed:
Push mapping alternative to test

hcorson-dosch-usgs committed:
Fix error w/ p4 fxn edit

hcorson-dosch-usgs committed:
revert to old extraction approach

hcorson-dosch-usgs committed:
Read in preds for eval sites instead of filtering all preds to eval sites

hcorson-dosch-usgs committed:
Add tar_assert statements

hcorson-dosch-usgs committed:
Sort by time as well

hcorson-dosch-usgs committed:
account for NAs in bias calc

hcorson-dosch-usgs committed:
Document # dates and #sites per bin

jordansread commented on 5_evaluate.R:
```diff
@@ -8,54 +8,51 @@ p5 <- list(
-  # And subset obs to those for sites w/ >= 10 dates with observations
-  ### Use https://github.com/USGS-R/mntoha-data-release/blob/main/src/eval_utils.R#L4-L11
+  # filter obs to sites and dates for which we have NLDAS output
+  # And further filter obs to those for sites w/ >= 10 dates with observations
```
Totally ticky-tacky, but since `min_obs_dates` is a parameter, this comment referring to 10 may get outmoded if the value changed. Seems pretty obvious to follow, but future-proofing the comment might look more like this:

`# And further filter obs to those for sites w/ >= `min_obs_dates` dates with observations`

jordansread commented on 5_evaluate.R:
```diff
@@ -8,54 +8,51 @@ p5 <- list(
+    {
+      tar_assert_identical(p5_eval_sites, unique(p5_obs_for_eval_groups$site_id), 
+                           "p5_eval_sites site id doesn't match p5_obs_for_eval_groups site id")
+      arrow::read_feather(sprintf('3_extract/out/GLM_%s_NLDAS.feather', p5_eval_sites)) %>%
```
Agreed that it makes sense to split up the reading and rely on a prior target here vs trying to read and filter all at once. This modularity of having `5_eval` rely on the general extracted outputs from `3_extract` is better for a variety of issues, including the one you mentioned about efficiency but also because we're making `5_eval` traceable for others - they can start with the same `3_` outputs and do their own filtering in the same way as us as opposed to relying on an `.nc` file that they wouldn't have access to.

jordansread commented on 5_evaluate.R:
```diff
@@ -65,19 +62,30 @@ p5 <- list(
-             match_pred_obs(eval_obs = p5_obs_for_eval_groups, eval_preds = p5_nldas_preds_eval_groups),
+             {
+               tar_assert_identical(unique(p5_obs_for_eval_groups$site_id), unique(p5_nldas_preds_eval_groups$site_id), 
+                                    "p5_obs_for_eval_groups site id doesn't match p5_nldas_preds_eval_groups site id")
```
👍

jordansread commented on 5_evaluate.R:
```diff
@@ -65,19 +62,30 @@ p5 <- list(
   # Add fields for year, doy, season, temp_bin
   tar_target(p5_nldas_pred_obs_surface_eval,
-             prep_data_for_eval(p5_nldas_pred_obs) 
+             prep_data_for_eval(p5_nldas_pred_obs, selected_depth = 0)
```
Like seasons and temperatures, I wonder if you need to use bins here instead of exact depth matches? Or provide the flexibility to do so?

hcorson-dosch-usgs replied:
Yes, agreed this isn't an ideal set up, given our plan to use depth bins. I think one option would be to remove the depth filtering from this step and instead add code to identify the depth class for each pred-obs pair. We could either use some [existing code](https://code.usgs.gov/wma/wp/lake-temperature-neural-networks/-/blob/master/3_assess/src/inspect_site.R#L81-93) to pull out only pred-obs at representative depths, or use each lake's depth to define depth bins (and thereby make use of all available pred-obs). Then we could add an additional argument and step to the plotting functions to filter the passed data to a specific depth bin before plotting.

hcorson-dosch-usgs replied:
Update - adding global (not lake-specific) depth bins for now -- e.g., all predictions <=1 meter are considered 'surface' predictions

jordansread commented on 5_evaluate.R:
```diff
@@ -93,45 +101,84 @@ p5 <- list(
-                          x_var = 'year', depth='surface', 
-                          outfile = '5_evaluation/out/nldas_surface_bias_year.png'),
+             plot_evaluation_barplot(p5_nldas_surface_bias_year, driver= 'NLDAS', 
+                          y_var = 'bias', x_var = 'year', depth='surface',
```
In the plots it would be helpful to know if bias is obs - model or model - obs. This could be done with the label. That makes the plots more interpretable.

jordansread commented on 5_evaluate.R:
```diff
@@ -93,45 +101,84 @@ p5 <- list(
   # Bar plot of bias through time, by doy
+  tar_target(p5_nldas_surface_bias_doy_png,
+             plot_evaluation_barplot(p5_nldas_surface_bias_doy, driver= 'NLDAS', 
+                          y_var = 'bias', x_var = 'doy', depth='surface',
```
Single DoY bars may not be as useful, since as you noted above, some dates have very few obs. I'd suggest binning DoY - we used a 3 day bin for Jared's paper, but you may want to make that flexible and perhaps more than 3, since we had a lot more obs in that paper.

jordansread commented on 5_evaluate.R:
```diff
@@ -93,45 +101,84 @@ p5 <- list(
   # Bar plot of bias by season
+  tar_target(p5_nldas_surface_bias_season_png,
+             plot_evaluation_barplot(p5_nldas_surface_bias_season, driver= 'NLDAS', 
+                          y_var = 'bias', x_var = 'season', depth='surface',
```
Does "depth" need to be consistent? I think above it is "0" (but perhaps could be a range) and here it is "surface".

hcorson-dosch-usgs replied:
Updated to term 'depth_class' and used in plot title (as was intended)

lindsayplatt commented on 5_evaluate.R:
```diff
@@ -8,54 +8,51 @@ p5 <- list(
+        select(-ice) %>%
+        arrange(site_id, time)
+    },
+    pattern = map(p5_eval_sites, p5_obs_for_eval_groups)
```
I think splitting this & the steps in `match_pred_obs()` is not necessary and is making the mapping more complicated. I think that you can actually join those together and skip some of these intermediate targets. After `p5_obs_for_eval_groups` and before `p5_nldas_pred_obs_csv`, you really only need one target (move some code from the target makefile into `match_pred_obs()` instead). It simplifies the mapping and eliminates any concern that we might not be using the same sites (meaning we don't need to use any of those `tar_asserts()`). 

```r
tar_target(p5_nldas_pred_obs, 
        match_pred_obs(
          file_in = sprintf("3_extract/out/GLM_%s_NLDAS.feather", unique(p5_obs_for_eval_groups$site_id)),
          site_obs_data = p5_obs_for_eval_groups
        ),
      pattern = map(p5_obs_for_eval_groups)
      )
    )
```

and `match_pred_obs()` would look like:

```r
match_pred_obs <- function(file_in, eval_obs) {
  eval_preds <- arrow::read_feather(file_in) %>%
    filter(time %in% eval_obs$time) %>%
    mutate(site_id = unique(eval_obs$site_id)) %>%
    select(-ice) %>%
    arrange(site_id, time)
  # The rest of what is already in `match_pred_obs()` goes here
}
```

You could go so far as to include the step to write out as a CSV within this step & then just have the next step `prep_data_for_eval()` read in a CSV.

hcorson-dosch-usgs replied:
Sounds good -- agreed that that would clean it up and removes risk re: not using the same sites. I'll make those edits now

hcorson-dosch-usgs replied:
Okay edits pushed. I think I landed in that awkward setup of targets after gradually shifting away from using the already read in predictions. Definitely seems cleaner to have this reduced # of targets now that we're simply reading in the predictions for each site.

lindsayplatt replied:
Yes, agree. Nice!

hcorson-dosch-usgs committed:
Move loading and filtering of eval preds into match_pred_obs()

jordansread commented on 5_evaluate/src/eval_utility_fxns.R:
```diff
@@ -1,24 +1,154 @@
+           doy = yday(time),
+           month = month(time),
+           season = case_when(
+             month >= 3 & month <= 5 ~ 'spring',
```
Often we use day of year/date to determine the seasonal bounds, which is different from the months (e.g., June 22nd instead of June 1 for start of summer). I'd recommend using day of year instead here.

jordansread commented on 5_evaluate/src/eval_utility_fxns.R:
```diff
@@ -1,24 +1,154 @@
+#' @param depth depth of the plotted pred-obs
+#' @param outfile The filepath for the exported png
+#' @return The filepath of the exported png 
+plot_evaluation_barplot <- function(plot_df, driver, y_var, x_var, depth, outfile) {
```
I think the bar plots are very appropriate for bias (basically "deviation from zero" makes sense w/ a bar). But for accuracy, bar plots are typically not used since zero isn't achievable by a model. So often accuracy would be a point, or a bound with distribution (box/violin). Not a necessary change for now, but something to consider for presentation of bias and accuracy and the things that do translate vs the small number of things that don't between the two.

jordansread commented on 5_evaluate/src/eval_utility_fxns.R:
```diff
@@ -1,24 +1,154 @@
+    rename(obs = temperature)
+  
+  # match up preds to test_obs, interpolating predictions to match the observation depths
+  pred_obs <- bind_rows(lapply(unique(eval_obs$time), function(selected_time) {
```
I don't think I can run this code w/o some significant work to get in here, but I've got some questions about this section. 

First, I think the warnings may get lost relative to which lake and/or observations are causing the error. Does this function get called per lake? Or for all data but grouped by lakes within the function? I think it is the former. You might want to clarify that in the documentation because I'm not sure based on the language "grouped by site" (could it be "filtered to one site" or "with column name {x} to be used to group by site"?)

I think this uses Alison's depth-matching code from [here](https://github.com/USGS-R/mntoha-data-release/blob/d94f73d3a09ddc6f43648b43f299529008ea9356/src/eval_utils.R#L113) which I'm good with. But we should note that Andy is currently binning observations in a different way (assigning the raw observations to nearest depth bin instead of interpolating). I think both approaches have their pros and cons (con of interp is the need for 2+ observations per date, **which is fairly limiting** since it will throw out observations that were the only one on a given day; con of binning is potential errors introduced in the bins are too coarse). 

My suggestion would be to create a new issue for how we bin/interp field measurements and once that decision is made it should be applied both here, and to what Andy is doing, (or he should pull depth-assigned observations from here or a data release that applies consistent code). So I'm good with keeping this as-is but want to flag the method for later consideration.

hcorson-dosch-usgs replied:
Yes - the function gets called per lake -- I've clarified that in the documentation. Given the mapping and parallelization on Tallgrass, no warnings came through in the console when I was running the pipeline, but I have modified the warnings/errors to include the site id -- thanks for catching that.

I've made a new issue - #59

jordansread reviewed (commented):
This looks good to me. Nice work implementing this as it gives us a good starting view of the model performance in different conditions. 

Feedback: 
1) I think most of the evaluation metrics need to do some kind of binning (depth _ranges_ as opposed to specific depths, DoY ranges vs specific days, and so on). So I think that kind of flexibility may need to be added
2) I flagged a topic for later consideration which is whether the observations should be binned to a depth or interpolated. Both have drawbacks. Currently the code interpolates, and I'm supportive of moving forward with this for the time being as swapping in a slightly different approach will only subtly impact the evaluation results.
3) Probably a simpler fix, but adding more info on bias (making it clear whether the model is hot or cold) would be helpful in the plots. Probably just an axis labeling fix.

hcorson-dosch-usgs committed:
Make minor edits suggested by Jordan

hcorson-dosch-usgs committed:
Add in depth bins - set globally

hcorson-dosch-usgs committed:
Fix merge conflicts from upstream

Merge branch 'main' of github.com:USGS-R/lake-temperature-process-models into NLDAS_surface_eval

# Conflicts:
#	1_prep.R

hcorson-dosch-usgs committed:
Add temp obs file to README, merge in upstream

hcorson-dosch-usgs committed:
Edit DOY metrics and plots to use DOY bins
